components/places/android/library/src/main/java/org/mozilla/places/PlacesConnection.kt

```diff
@@ -0,0 +1,161 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.places
+
+import com.sun.jna.Pointer
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.Closeable
+
+open class PlacesConnection(path: String, encryption_key: String? = null) : Closeable {
+
+    private var db: RawPlacesConnection? = null
+
+    init {
+        this.db = rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_connection_new(path, encryption_key, error)
+        }
+    }
+
+    @Synchronized
+    override fun close() {
+        val db = this.db
+        this.db = null
+        if (db != null) {
+            LibPlacesFFI.INSTANCE.places_connection_destroy(db)
+        }
+    }
+
+    // This is actually handled internally by kotlin, apparently. Not sure
+    // why it doesn't need an `override` but it doesn't.
+    @Suppress("Unused")
+    fun finalize() {
+        this.close()
+    }
+
+    fun noteObservation(data: VisitObservation) {
+        val json = data.toJSON().toString()
+        rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_note_observation(this.db!!, json, error)
+        }
+    }
+
+    fun queryAutocomplete(query: String, limit: Int = 10): List<SearchResult> {
+        val json = rustCallForString { error ->
+            LibPlacesFFI.INSTANCE.places_query_autocomplete(this.db!!, query, limit, error)
+        }
+        return SearchResult.fromJSONArray(json)
+    }
+
+    fun getVisited(urls: List<String>): List<Boolean> {
+        val urlsToJson = JSONArray()
+        for (url in urls) {
+            urlsToJson.put(url)
+        }
+        val urlStr = urls.toString()
+        val visitedStr = rustCallForString { error ->
+            LibPlacesFFI.INSTANCE.places_get_visited(this.db!!, urlStr, error)
+        }
+        val visited = JSONArray(visitedStr)
+        val result = mutableListOf<Boolean>()
+        for (index in 0 until visited.length()) {
+            result.add(visited.getBoolean(index))
+        }
+        return result
+    }
+
+    @Synchronized
+    private fun <U> rustCall(callback: (RustError.ByReference) -> U): U {
+        val e = RustError.ByReference()
+        // TODO: should we synchronize on an Object instead?
```

Comment:
As in `synchronized(this)`? Also, out of curiosity, is the locking a requirement of JNA, or our implementation? I'm guessing our implementation...if not, do we need to hold the lock when we call `places_connection_destroy` or `places_destroy_string`? Could `close` be called concurrently with any of the other methods?

Reply:
As in keep a private local object instead of using Synchronized. That said, the way we do this in the recently-landed FxA patch is better (Sadly, code reuse here is hard due to having different Error types, which need to be different due to having different functions that need to be called to release Strings. Eventually we'll probably untangle this...)